### PR TITLE
`Development`: Extract build agent git service

### PR DESCRIPTION
**Pull request by krusche**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

Currently, there is an inconsistency in build agents when using single and multi-node setup. In multi-node, they use https. In single-node, they copy the folder. This PR fixes this issue and has the added benefit of reducing the responsibilities of `GitService`

### Description
This PR implements several structural and code improvements:
* Separate `GitService` and `BuildJobGitService` and added common functionality in `AbstractGitService`
* Adjust the rest of code and tests
* Change config files so that build agents have the necessary credentials to clone repos from localvc
* Reduce the amount of services used in build agents and avoid error messages when build agents are started without a cluster
* Rename and move some classes and packages for more clarity
* Improve jgit server code

### Steps for Testing

> [!NOTE]
> **Test on both setups! (single node and multi node servers with ICL)**

Prerequisites:
- 1 Students
- 1 Programming Exercise

1. Submit exercise and make sure that build runs
2. Clone repository and make sure that it's cloned correctly
3. Play around with features related to localvc/ci

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java [hunk 1]
```diff
@@ -1404,6 +1152,7 @@ public void combineAllCommitsIntoInitialCommit(Repository repo) throws IllegalSt
      * @throws IOException if the deletion of the repository failed.
      */
     public void deleteLocalRepository(Repository repository) throws IOException {
+        super.deleteLocalRepository(repository);
```
Comment (BBesrour) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java:1155]: We can remove the rest of the method and only keep `cachedRepositories.remove(repoPath);`

File: src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractGitService.java [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+                transport.setTimeout(JGIT_TIMEOUT_IN_SECONDS);
+                sshTransport.setSshSessionFactory(sshSessionFactory);
+            }
+            else {
+                log.error("Cannot use ssh properly because of mismatch of Jgit transport object: {}", transport);
+            }
+        };
+    }
+
+    protected static SshdSessionFactoryBuilder getSshdSessionFactoryBuilder(Optional<String> gitSshPrivateKeyPath, Optional<String> gitSshPrivateKeyPassphrase, URL gitUrl) {
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractGitService.java:105]: Can you extract the inner classes from this method? It's quite hard to follow which method belongs to what

File: src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractGitService.java [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+     *                                     <li>https://stackoverflow.com/questions/45266021/java-jgit-files-delete-fails-to-delete-a-file-but-file-delete-succeeds</li>
+     *                                     <li>https://git-scm.com/docs/git-gc</li>
+     *                                     <li>https://www.eclipse.org/lists/jgit-dev/msg03734.html</li>
+     *                                     </ul>
+     *                                     </p>
+     */
+    @NotNull
+    public static Repository linkRepositoryForExistingGit(Path localPath, VcsRepositoryUri remoteRepositoryUri, String defaultBranch, boolean isBare)
+            throws IOException, InvalidRefNameException {
+        // Create the JGit repository object
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractGitService.java:226]: I think this is a left over?

File: src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractGitService.java [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+
+        // Open the repository from the filesystem
+        FileRepositoryBuilder builder = new FileRepositoryBuilder();
+
+        if (isBare) {
+            builder.setBare();
+        }
+        // bare repositories use a different path than checked out ones
+        builder.setGitDir(isBare ? localPath.toFile() : localPath.resolve(".git").toFile()).setInitialBranch(defaultBranch).setMustExist(true).readEnvironment().findGitDir()
+                .setup(); // scan environment GIT_* variables
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractGitService.java:236]:
```suggestion
        if (isBare) {
            builder.setBare();
            builder.setGitDir(localPath.toFile());
        }
        else {
            builder.setGitDir(localPath.resolve(".git").toFile());
        }
        builder.setInitialBranch(defaultBranch).setMustExist(true).readEnvironment().findGitDir().setup(); // scan environment GIT_* variables
```

File: src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractGitService.java [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+
+        try (Repository repository = new Repository(builder, localPath, remoteRepositoryUri)) {
+            // disable auto garbage collection because it can lead to problems (especially with deleting local repositories)
+            // see https://stackoverflow.com/questions/45266021/java-jgit-files-delete-fails-to-delete-a-file-but-file-delete-succeeds
+            // and https://git-scm.com/docs/git-gc for an explanation of the parameter
+            // and https://www.eclipse.org/lists/jgit-dev/msg03734.html
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractGitService.java:242]: These links are already in the javadoc of this method. Do we want to repeat them again?

File: src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractGitService.java [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+        if (repoUri == null || repoUri.getURI() == null) {
+            return null;
+        }
+        // Get HEAD ref of repo without cloning it locally
+        try {
+            log.debug("getLastCommitHash {}", repoUri);
+            var headRef = lsRemoteCommand().setRemote(getGitUriAsString(repoUri)).callAsMap().get(Constants.HEAD);
+
+            if (headRef == null) {
+                return null;
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractGitService.java:291]: Why does this not throw an exception?

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/BuildJobGitService.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+            CredentialsProvider.setDefault(new UsernamePasswordCredentialsProvider(gitUser, gitToken.get()));
+        }
+        else {
+            log.info("BuildJobGitService will use username + password as authentication method to interact with remote git repositories");
+            CredentialsProvider.setDefault(new UsernamePasswordCredentialsProvider(gitUser, gitPassword));
+        }
+    }
+
+    private void configureSsh() {
+
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/BuildJobGitService.java:66]:
```suggestion
```

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

**Review by pzdr7 — APPROVED** [review]
Tested on TS1.

Things I've tried:

- Create programming exercise ✅ 
- Trigger template, solution builds ✅ got a result for each
- Make submission using code editor ✅  got a result
- Make submission using cloned repo ✅  got a result
- Cancel a build job ✅

File: src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractGitService.java (hunk 6 above)
Comment (BBesrour) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractGitService.java:209]: Since the PR focuses on separating concerns for GitService, we want to keep the same behavior for now. Changing this would modify behavior of some other parts which would out of scope of this PR.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/BuildJobExecutionService.java [hunk 8]
```diff
@@ -450,15 +450,18 @@ private Path cloneRepository(VcsRepositoryUri repositoryUri, @Nullable String co
         try {
             // Clone the assignment repository into a temporary directory
             // TODO: use a random value if commitHash is null
-            Repository repository = gitService.getOrCheckoutRepository(repositoryUri, Paths.get(CHECKED_OUT_REPOS_TEMP_DIR, commitHash, repositoryUri.folderNameForRepositoryUri()),
-                    false);
+            Repository repository = buildJobGitService.cloneRepository(repositoryUri,
+                    Paths.get(CHECKED_OUT_REPOS_TEMP_DIR, commitHash, repositoryUri.folderNameForRepositoryUri()));
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/BuildJobExecutionService.java:454]: (nitpick)
we prefer `Path.of` over `Paths.get`.
```suggestion
                    Path.of(CHECKED_OUT_REPOS_TEMP_DIR, commitHash, repositoryUri.folderNameForRepositoryUri()));
```

File: src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractGitService.java [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+    }
+
+    protected static SshdSessionFactoryBuilder getSshdSessionFactoryBuilder(Optional<String> gitSshPrivateKeyPath, Optional<String> gitSshPrivateKeyPassphrase, URL gitUrl) {
+        var credentialsProvider = new CustomCredentialsProvider();
+
+        CredentialsProvider.setDefault(credentialsProvider);
+
+        return new SshdSessionFactoryBuilder().setKeyPasswordProvider(keyPasswordProvider -> new CustomKeyPasswordProvider(gitSshPrivateKeyPath, gitSshPrivateKeyPassphrase))
+                .setConfigStoreFactory((homeDir, configFile, localUserName) -> new CustomSshConfigStore(gitUrl))
+                .setSshDirectory(new java.io.File(gitSshPrivateKeyPath.orElseThrow())).setHomeDirectory(new java.io.File(System.getProperty("user.home")));
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractGitService.java:112]: Can you import the `File` class instead of using full qualified names?

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/BuildJobGitService.java [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+@Profile(PROFILE_BUILDAGENT)
+@Service
+public class BuildJobGitService extends AbstractGitService {
+
+    private static final Logger log = LoggerFactory.getLogger(BuildJobGitService.class);
+
+    public BuildJobGitService() {
+        super();
+    }
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/BuildJobGitService.java:40]:
```suggestion
```

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/BuildJobGitService.java [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+     * @throws URISyntaxException   if the URI is invalid
+     */
+    public Repository cloneRepository(VcsRepositoryUri repoUri, Path localPath) throws GitAPIException, GitException, InvalidPathException, IOException, URISyntaxException {
+        var gitUriAsString = getGitUriAsString(repoUri);
+        log.debug("Cloning from {} to {}", gitUriAsString, localPath);
+        // make sure the directory to copy into is empty (the operation only executes a delete if the directory exists)
+        FileUtils.deleteDirectory(localPath.toFile());
+        Git git = cloneCommand().setURI(gitUriAsString).setDirectory(localPath.toFile()).call();
+        git.close();
+        return getExistingCheckedOutRepositoryByLocalPath(localPath, repoUri, defaultBranch);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/BuildJobGitService.java:132]:
```suggestion
        try (Git git = cloneCommand().setURI(gitUriAsString).setDirectory(localPath.toFile()).call()) {
            return getExistingCheckedOutRepositoryByLocalPath(localPath, repoUri, defaultBranch);
        }
```

**Review by az108 — DISMISSED** [review]
Works as expected, tested in testing session together and tested it myself aswell

**Review by SimonEntholzer — DISMISSED** [review]
Tested in testing session, and code (mostly small changes) look good

**Review by pzdr7 — DISMISSED** [review]
Reapprove

**Review by Jan-Thurner — COMMENTED** [review]
The code looks good for the most part. However, I have a few questions in the comments.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractGitService.java [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+        repository.closeBeforeDelete();
+        FileUtils.deleteDirectory(repoPath.toFile());
+        log.debug("Deleted Repository at {}", repoPath);
+    }
+
+    static class CustomCredentialsProvider extends CredentialsProvider {
+
+        @Override
+        public boolean isInteractive() {
+            return false;
```
[the remark this answers is not in the document]
Comment (BBesrour) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractGitService.java:261]: This was taken from gitService and we shouldn't change it in this PR since we want to keep the same behavior (mostly).

**Review by Jan-Thurner — DISMISSED** [review]
Code LGTM

**Review by SimonEntholzer — DISMISSED** [review]
Tested on Ts3

**Review by pzdr7 — APPROVED** [review]
Tested again on TS3.

- Created a programming exercise ✅ 
- Triggered template / solution builds manually; got a result each time ✅ 
- Participated; got a result ✅ 
- Canceled a build job ✅

**Review by BBesrour — APPROVED** [review]
Tested on TS2 and staging

**Review by pzdr7 — APPROVED** [review]
Tested on LocalCI Staging
